app/pipes.go

```diff
@@ -0,0 +1,174 @@
+package main
+
+import (
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/gorilla/mux"
+
+	"github.com/weaveworks/scope/xfer"
+)
+
+const (
+	gcInterval  = 30 * time.Second // we check all the pipes every 30s
+	pipeTimeout = 1 * time.Minute  // pipes are closed when a client hasn't been connected for 1 minute
+	gcTimeout   = 2 * time.Minute  // after 2 minutes, tombstoned pipes are forgotten
+)
+
+type pipeRouter struct {
+	sync.Mutex
+	wait    sync.WaitGroup
+	quit    chan struct{}
+	pipes   map[string]xfer.Pipe
+	lastPut map[string]time.Time // for timeing out pipes
+}
+
+func registerPipeRoutes(router *mux.Router) *pipeRouter {
+	pipeRouter := &pipeRouter{
+		quit:    make(chan struct{}),
+		pipes:   map[string]xfer.Pipe{},
+		lastPut: map[string]time.Time{},
+	}
+	pipeRouter.wait.Add(1)
+	go pipeRouter.gcLoop()
+	router.Methods("GET").Path("/api/pipe/{pipeID}").HandlerFunc(pipeRouter.handleUIPipeWS)
+	router.Methods("DELETE").Path("/api/pipe/{pipeID}").HandlerFunc(pipeRouter.deletePipe)
+	router.Methods("GET").Path("/api/pipe/{pipeID}/probe").HandlerFunc(pipeRouter.handleProbePipeWS)
+	return pipeRouter
+}
+
+func (pr *pipeRouter) stop() {
+	close(pr.quit)
+	pr.wait.Wait()
+}
+
+func (pr *pipeRouter) getOrCreatePipe(id string, track bool) (xfer.Pipe, bool) {
+	pr.Lock()
+	defer pr.Unlock()
+	pipe, ok := pr.pipes[id]
+	if !ok {
+		log.Printf("Creating pipe id %s", id)
+		pipe = xfer.NewPipe(id)
+		pr.pipes[id] = pipe
+	}
+	if pipe.Closed() {
+		return nil, false
+	}
+	if track {
+		delete(pr.lastPut, id)
+	}
+	return pipe, true
+}
+
+func (pr *pipeRouter) gcLoop() {
+	defer pr.wait.Done()
+	ticker := time.Tick(gcInterval)
+	for {
+		select {
+		case <-pr.quit:
+			return
+		case <-ticker:
+		}
+
+		pr.timeoutPipes()
+		pr.gcPipes()
+	}
+}
+
+func (pr *pipeRouter) timeoutPipes() {
+	pr.Lock()
+	defer pr.Unlock()
+	now := time.Now()
+	for id, lastPut := range pr.lastPut {
+		if now.Sub(lastPut) > pipeTimeout {
+			log.Printf("Timeing out pipe %s", id)
+			delete(pr.lastPut, id)
+			pipe := pr.pipes[id]
+			pipe.Close(now)
+		}
+	}
+}
+
+func (pr *pipeRouter) gcPipes() {
+	pr.Lock()
+	defer pr.Unlock()
+	now := time.Now()
+	for pipeID, pipe := range pr.pipes {
+		if pipe.Closed() && now.Sub(pipe.ClosedAt()) > gcTimeout {
+			log.Printf("GCing pipe %s", pipeID)
+			delete(pr.pipes, pipeID)
+		}
+	}
+}
+
+func (pr *pipeRouter) handleUIPipeWS(w http.ResponseWriter, r *http.Request) {
```

Comment:
Yes would be nice, but I can't think of a way that isn't more horrific than having two similar functions.  The key differences are (1) they talk to different ends of the pipe and (2) one does reference counting to trigger pipe timeouts.